draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1581,75 +1581,143 @@ Steps:
+For both of these curves, {{RFC7748}} lists both the Montgomery and twisted Edwards
+forms and gives the corresponding rational maps.
+
+The birational map for edwards25519 ({{RFC7748}}, Section 4.1)
+uses the constant sqrt\_neg\_486664 = sqrt(-486664) mod 2^255 - 19.
+To ensure compatibility, this constant MUST be chosen such that
+sgn0(sqrt\_neg\_486664) == 1.
+Analogous ambiguities in other standardized birational maps MUST be
+resolved in the same way: for any constant k whose sign is ambiguous,
+k MUST be chosen such that sgn0(k) == 1.
+
+The 4-isogeny map from curve448 to edwards448 ({{RFC7748}}, Section 4.2)
+is unambiguous with respect to sign.
+
+When defining new twisted Edwards curves, a Montgomery equivalent and birational
+map SHOULD be specified, and the sign of the birational map SHOULD be stated
+unambiguously.
+
+When hashing to a twisted Edwards curve that does not have a standardized
+Montgomery form or birational map, the following procedure MUST be
+used to derive them.
 For a twisted Edwards curve given by a * x^2 + y^2 = 1 + d * x^2 * y^2,
-first compute A and B, the parameters of the equivalent Montgomery curve,
-as follows:
+first compute A and B, the parameters of the equivalent Montgomery curve
+y'^2 = x'^3 + A * x'^2 + B * x', as follows:
 
-- A = (a + 2) / 2
+- A = (a + d) / 2
 - B = (a - d)^2 / 16
 
-Next, use A and B as the curve parameters in the Elligator 2 method of
-{{elligator2}} to obtain a point (x', y') on the Montgomery curve.
-Finally, convert (x', y') to a point (x, y) on the target curve.
-Letting B' = 4 / (a - d), compute
+Next, let B' = 4 / (a - d). Then the birational map from the point (x', y')
+on the Montgomery curve to the point (x, y) on the twisted Edwards curve is
+given by
 
 - x = x' / y'
 - y = (B' * x' - 1) / (B' * x' + 1)
 
-This can be done in one inversion using Montgomery's trick {{M87}}:
-invert the product y' * (B' * x' + 1), then multiply by y' to obtain
-1 / (B' * x' + 1), and likewise for 1 / y'.
+For completeness, we give the inverse map in {{birational-map-inverse}}.
+Note that the inverse map is not used when hashing to a twisted Edwards curve.
+
+Note that the above map can be evaluated with just one inversion using
+Montgomery's trick {{M87}}.
+First, compute t = y' * (B' * x' + 1).
+Then t * y' == 1 / (B' * x' + 1) and t * (B' * x' + 1) == 1 / y'.
+
+Birational maps are undefined when the denominator of either
+rational function is zero.
+For example, in the map described above, the exceptional cases are
+y' == 0 or B' * x' == -1.
+Implementations MUST detect exceptional cases and return the value
+(x, y) = (0, 1), which is a valid point on all twisted Edwards curves
+given by the equation above.
+
+The following straight-line implementation of the above rational map
+handles the exceptional cases.
+Implementations of other rational maps (e.g., the ones give in {{RFC7748}})
+are analogous.
+
+~~~
```

Comment:
Consider to refactor as follows:
```
map_to_curve_elligator2_edwards(u)
Input: u, an element of F.
Output: (x, y), a point on E.

1. (x1, y1) = map_to_curve_elligator2(u)    // (x1, y1) is on M
2. 1. t1 = y' * B'
3. t2 = x' + 1
4. t3 = t1 * t2
5. t3 = inv0(t3)
6.  x = t2 * t3
7.  x = x * x'
8.  y = x' - 1
9.  y = y * t3
10.  y = y * t1
11. e = y == 0
12. y = CMOV(y, 1, e)
13. return (x, y)
```

Reply:
I think I don't get it. The point of this PR is to break apart the definition of `birational_map` from the definition of `map_to_curve_elligator2_edwards`, and the above suggestion puts them back together. This is a problem because `birational_map` for existing curves is *not* the above map---see RFC7748, for example.

Reply:
Or is the suggestion changing x' -> x1, and y' -> y1?

Reply:
I want to understand all of this.

Following the description of the document. It says that curve25519 and curve448 have special birational mappings and an implementor must follow them. Then, the document says that for all other cases, it should follow the `bitarional_map` you gave. 

So let me rephrase in pseudo-code:
```
map_to_curve_elligator2_edwards(u)
(x1,y1) = elligator2(u)
 if curve == {edwards25519, edwards448}
    use map \phi from RFC7748
else
   use map \phi explicitly given as .....
```
Is this the recommendation given?

Reply:
Not quite. The document says that *any* curve that already has a birational map specified in a standard should use that map. The examples we give are '25519 and '448, but it seems extremely likely that any Montgomery curve that gets standardized will end up standardizing an Edwards curve, too: if a curve is good enough for key exchange, it's good enough for EdDSA.

So the point is, if there's already a standardized birational map, use that one so that you get "free" compatibility between the Edwards and the Montgomery hashing. As you pointed out above, this might break if we add a ciphersuite string---to me, this is all the more reason not to add one.